### ISSUE TITLE
Unexpected InvalidReturnStatement Error.

### ISSUE THREAD

**Issue by azjezz** [vimeo/psalm#2152]
I encountered this issue with a basic `abs()` implementation : 

i can't tell if this is expected or a bug.

code ( https://psalm.dev/r/d9a3a763c7 ) : 
```php
<?php

namespace Math;
  
/**
 * Return the absolute value of the given number.
 *
 * @psalm-template T as numeric
 *
 * @psalm-param T $number
 *
 * @psalm-return T
 */
function abs($number)
{
    return $number < 0 ? -$number : $number;
}
```

Expected output : 
```console
No issues!
```

Actual output : 

```console
ERROR: InvalidReturnStatement - 27:12 - The type 'int|T as numeric' does not match the declared return type 'T as numeric' for Math\abs

ERROR: InvalidReturnType - 23:18 - The declared return type 'T' for Math\abs is incorrect, got 'int|T'
```

when trying to do `$number*-1` instead of `-$number` ( https://psalm.dev/r/99723ecd30 ) :
 
```console
INFO: MixedOperand - 27:26 - Left operand cannot be mixed
```

**Comment by muglug** [comment]
Definitely a bug, although that should read `T as int|float` to avoid allowing `"5"` which counts as numeric to PHP.

**Comment by azjezz** [comment]
Ah, thanks :) i thought `numeric` is equivalent to hack's `num`. maybe psalm can add a `num` or `number` union for `int|float` ?

**Comment by muglug** [comment]
at the moment I don't want to introduce another reserved word in the type system if I can help it - Hack has the luxury of being able to control those completely.

**Comment by azjezz** [comment]
i have found another issue that might be related to `$number*-1` : https://psalm.dev/r/4479ce864c

**Comment by weirdan** [comment]
> I have found another issue that might be related to `$number*-1` : https://psalm.dev/r/4479ce864c

Doesn't seem to be related to multiplication. 

Simplified: https://psalm.dev/r/7f5efcfa77
```php
<?php
/**
 * @template T as int
 * @param T $i
 */
function f($i): float
{
  return atan($i + $i);
}
```
Expected: no issues
Actual:
```
Psalm output (using commit f6b0de0):

INFO: MixedOperand - 8:15 - Left operand cannot be mixed

INFO: MixedArgument - 8:15 - Argument 1 of atan cannot be mixed, expecting float
```

**Comment by orklah** [comment]
This is basically the same issue than https://github.com/vimeo/psalm/issues/6380. I explained how to resolve there so I'm closing this one
